amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala

```diff
@@ -191,26 +191,31 @@ final case class AmqpCachedConnectionProvider(provider: AmqpConnectionProvider,
         state.compareAndSet(Connecting, Connected(connection, 1))
         connection
       } else get
+    case Connecting => get
     case c @ Connected(connection, clients) =>
       if (state.compareAndSet(c, Connected(connection, clients + 1))) connection
       else get
-    case _ => get
+    case Closing => get
+    case _ => throw new RuntimeException("Unknown provider state.")
   }
 
   @tailrec
   override def release(connection: Connection): Unit = state.get match {
+    case Empty => ()
+    case Connecting => release(connection)
     case c @ Connected(cachedConnection, clients) =>
       if (cachedConnection != connection)
         throw new IllegalArgumentException("Can't release a connection that's not owned by this provider")
 
-      if (state.compareAndSet(c, Closing)) {
-        if (clients == 1 || !automaticRelease) {
+      if (clients == 1 || !automaticRelease) {
+        if (state.compareAndSet(c, Closing)) {
           provider.release(connection)
           state.compareAndSet(Closing, Empty)
-        } else {
-          state.compareAndSet(Closing, Connected(cachedConnection, clients - 1))
         }
```

Comment:
If we lost this cas to get to close, we need to recurse

Reply:
You mean `state.compareAndSet(Closing, Empty)`?

Didn't you say above that you rather throwing an exception since we should never lose it?

Reply:
I think GitHub made something weird with the offset, because now I'm not sure which one it relates to. I'll take another look at the complete source.

Reply:
In general for *every* cas you do, you should look at the return value and act in some way. I think you are correct here, if it was changed from `Closing` to something else by another thread that is a bug, so throw IllegalState for that.